New reserveUploadToPaths Protocol; Refactor DataLayer Classes; UPath

fm3 opened the pull request:
### Description

This PR changes how clients with direct access to the underlying data storage can get datasets registered into WK.

Corresponding libs client PR: https://github.com/scalableminds/webknossos-libs/pull/1359

<details>
<summary>New ReserveDatasetUploadToPaths Protocol</summary>

 - Client sends reserveDatasetUploadToPaths request to wk side, including a preliminary datasource plus layersToLink.
 - Server generates id and paths for all mags and attachments where the client must put the stuff. This is returned to the client.
 - Server also validates layersToLink, and creates a dataset in the database, marked as not yet fully uploaded to the target paths.
 - After the client has actually put the data there, it calls finishDatasetUploadToPath, marking the dataset as complete. It can now be used.
 - Note that layersToLink here are by datasetId+layerName, but for normal dataset upload it’s still datasetDirectoryName+layerName (we’ll change that to id later)

</details>

<details>
<summary>New ReserveAttachmentUploadToPath Protocol</summary>

 - Client sends reserveAttachmentUploadToPath request with dataset id, attachment name, type etc. WK returns the path where the client must put the attachment. Wk inserts this path in the DB, marked as not yet fully uploaded
 - After the client has actually put the data there, it calls finishAttachmentUploadToPath with the same parameters (as there is no attachment id), marking the attachment as complete. It can now be used.

</details>

<details>
<summary>New ReserveUploadToPathForPreliminary Protocol</summary>
This extra route supports a new workflow, which will be used only by the convert_to_wkw worker job, which webknossos starts when a user uploads a dataset that is not yet in wkw/zarr format. When starting this upload, a dataset is already inserted in the database (via reserveUpload – not manual). Then after finishUpload, the conversion worker job is started. Now after the worker has converted the dataset to wkw/zarr, it needs to call this new reserveUploadToPathsForPreliminary route, which works only on datasets in this particular state. It returns paths, just like the normal, user-facing reserveUploadToPaths described above. But it is different in that it needs to operate on the datasetId of the already-created, preliminary dataset from the original reserveUpload.

</details>

<details>
<summary>Changes to Editing DataSources</summary>

 - Dataset settings view no longer allows changing layer names
 - It also no longer allows editing the json directly (textarea is hidden for now, to be removed in follow-up)
 - The backend now only applies the allowed changes on save. (category, boundingBox, coordinatesTransformations, defaultViewConfiguration, adminViewConfiguration, largestSegmentId, voxelSize, deleted layers)

</details>

<details>
<summary>Misc Changes</summary>

 - localDirectoryWhitelist is no longer checked in data loading, only in add/exploreAndAdd/upload
 - exploreAndAdd now returns dataset id
 - full dataSource now included in dataset publicWrites, with source data paths if requested by includePaths

</details>

<details>
<summary>Refactoring: UPath</summary>

 - New class UPath handles paths that can be either remote (s3, https, gcs) or local (file or no schem)
 - Attachments and mags have this UPath class now instead of URI and String respectively
 - VaultPath also wraps a UPath (plus its DataVault with credentials)

</details>

<details>
<summary>Refactoring: New DataLayer Class Hierarchy</summary>

 - Simplified DataSource and DataLayer classes.
<img width="800" alt="image" src="https://github.com/user-attachments/assets/4de47da2-ffae-4149-8e70-4653ca86d630" />

</details>

### Steps to test:
<details>
<summary>Test against regressions</summary>

- Browse some data, annotate some, explore remote dataset, explore local dataset in directoryWhitelist, should all still work
- Test that both the newest and also older libs versions can still open_remote and list existing datasets (including ones with unusual data, like remote N5)

</details>

<details>
<summary>Test new features</summary>

- With new libs client code, go through the reserveUploadToPaths flow https://github.com/scalableminds/webknossos-libs/pull/1359
- Also test reserveAttachmentUploadToPath via libs
- Test that libs can still upload a dataset.
- Attempt to change paths of an existing dataset in wk, or add a dataset with paths outside of itself, should not be possible (security!)
- Change the application.conf `webKnossos.datasets.uploadToPathsPrefixes` to our managed s3 object storage, give both wk and the client credentials, should also work with this.

</details>

### TODOs:
<details>
<summary>Backend</summary>

- [x] Decide how the directory name should look: just the id? or prefix-id? (might get outdated on renamings)
- [x] is the reserve response the full datasource again? or just a path prefix?
- [x] how to configure this in the dev setup
- [x] clean up the whole DataSource mess?
  - [x] ensure we include resolutions;
  - [x] never leak credentialIds (except during explore/add flow);
    - [x] settings view reads InboxDataSource, which includes credentialIds.
    - [x] Needs new Api for changing relevant fields
  - [x] compact json version? (resolutions, no paths?)
  - [x] do not write resolutions to disk?
- [x] directory scan overwrites status to No longer available on the datastore
- [x] reserveAttachmentUploadToPaths
- [x] what’s up with the zarr streaming snapshots?
- [x] evolutions
- [x] does upload still check unique directory name?
- [x] ensure that add route still checks localDirectoryWhiteList (also for exploreAndAdd)
- [x] move stuff to DatasetService, make createDataset private again
- [x] re-test upload with layersToLink from libs
- [x] return readable failures on now-unsupported routes? (old datastore-based reserveManualUpload)
- [x] do not require status field in datasource in reserveUpload request body.
- [x] introduce api version 11
- [x] re-test that editing DS in DB + reload propagates to datastore
- [x] on finishAttachmentUploadToPath, update datasource on disk too if isVirtual=False.
- [x] datasource update call via wk backend instead of frontend->datastore? skip datastore if isVirtual=true
- [x] more unit tests for UPath
- [x] double check that dataset info route is backwards compatible for libs
- [x] add remote UI layer name no longer editable, but explore returns empty, that’s a problem. test with first entry from notion remote datasets table.

</details>
<details>
<summary>Frontend</summary>

- [x] dataset settings view: include paths, remove advanced view
- [x] updating datasource should go via wk first

</details>

### Follow-Ups
<details>
<summary>Possible Follow-Up Issues</summary>

 - #8939
 - #8940
 - #8941
 - #8942
 - #8943
 - #8944
 - #8945
 - #8946

</details>

### Issues:

- fixes #8827
- fixes #8859
- fixes #8857 
- fixes #8851
- fixes #8762
- fixes #8881 
- fixes #8929
- contributes to #8749


------
- [x] Added changelog entry (create a `$PR_NUMBER.md` file in `unreleased_changes` or use `./tools/create-changelog-entry.py`)
- [x] Added migration guide entry if applicable (edit the same file as for the changelog)
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change https://github.com/scalableminds/webknossos-libs/pull/1359
- [x] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [x] Needs datastore update after deployment

MichaelBuessemeyer commented on test/backend/DataSourceTestSuite.scala:
```diff
@@ -59,10 +63,12 @@ class DataSourceTestSuite extends PlaySpec {
+              DataLayerAttachments(
+                meshes = Seq(LayerAttachment("meshfile",
+                                             UPath.fromStringUnsafe("./meshes/meshfile"),
+                                             LayerAttachmentDataformat.zarr3))))
```
could you please add an inequality check below?

Kinda alike
``` 
val notEqualDataSource = dataSourceCopyPastedFromAbove.copy(name="testLayerChanged")
assert(dataSource.hashCode() != notEqualDataSource.hashCode())
```
maybe?

fm3 replied:
Good idea, will do!

MichaelBuessemeyer commented on test/backend/DataVaultTestSuite.scala:
```diff
@@ -228,49 +230,60 @@ class DataVaultTestSuite extends PlaySpec {
+        "correctly identify equality" in {
+          val mockDataVault = new MockDataVault
+          val somePath = new VaultPath(UPath.fromStringUnsafe("protocol://host/a/b"), mockDataVault)
+          val someOtherPath = new VaultPath(UPath.fromStringUnsafe("protocol://host/a/b"), mockDataVault)
```
Maybe `someOtherPath => samePath` would be a better name here?

MichaelBuessemeyer commented on test/backend/DataVaultTestSuite.scala:
```diff
@@ -228,49 +230,60 @@ class DataVaultTestSuite extends PlaySpec {
+        "correctly identify inequality" in {
+          val mockDataVault = new MockDataVault
+          val somePath = new VaultPath(UPath.fromStringUnsafe("protocol://host/a/b"), mockDataVault)
+          val someOtherPath = new VaultPath(UPath.fromStringUnsafe("protocol://host/a/c"), mockDataVault)
```
here `someOtherPath` makes more sense imo

MichaelBuessemeyer commented on test/backend/UPathTestSuite.scala:
```diff
@@ -0,0 +1,130 @@
+      assert(UPath.fromString("/absolute/some$where").exists(_.toString == "/absolute/some$where"))
+      assert(UPath.fromString("/absolute/with²Unicode").exists(_.toString == "/absolute/with²Unicode"))
+      assert(UPath.fromString("/absolute/ with space").exists(_.toString == "/absolute/ with space"))
+      assert(UPath.fromString("file:///absolute/with²Unicode").exists(_.toString == "/absolute/with²Unicode"))
```
So toString does not print the scheme? This might get confusing in the debug / server logs when the schem is missing :thinking:?

Edit: Oh just saw below: So this only goes for local files / paths? In that case, this should be fine I think

fm3 replied:
Exactly, local paths will always be printed starting with `./` or `/` (but parsing is more permissive). Remote paths will always be printed with scheme.

MichaelBuessemeyer commented on test/backend/UPathTestSuite.scala:
```diff
@@ -0,0 +1,130 @@
+
+    "resolve strings correctly (remote)" in {
+      assert(
+        (UPath.fromStringUnsafe("https://hello.com/there") / "subkey").toString == "https://hello.com/there/subkey"
```
![](https://www.colorado.edu/linguistics/sites/default/files/styles/medium_750px_50_display_size_/public/article-image/kenobi-thumb.png?itok=aLM6EHnI)
```suggestion
        (UPath.fromStringUnsafe("https://hello.com/there") / "generalKenobi").toString == "https://hello.com/there/generalKenobi"
```

MichaelBuessemeyer commented on app/models/dataset/DatasetService.scala:
```diff
@@ -277,11 +256,112 @@ class DatasetService @Inject()(organizationDAO: OrganizationDAO,
+              u.adminViewConfiguration,
+              u.coordinateTransformations,
+              e.additionalAxes,
+              e.attachments,
```
Oh nvm.There is `reserveManualAttachmentUpload`

MichaelBuessemeyer commented on unreleased_changes/8844.md:
```diff
@@ -0,0 +1,14 @@
+- Introduced HTTP API version 11
+- The API for the python client for reserving manual uploads (in case the python client has access to the underlying storage) has changed. WK now decides the storage paths.
+- There is now an API for the python client for reserving manual dataset attachment uploads (in case the python client has access to the underlying storage).
+- Refactored the DataLayer class hierarchy.
```
and DataSource?

MichaelBuessemeyer commented on unreleased_changes/8844.md:
```diff
@@ -0,0 +1,14 @@
+
+### Breaking Changes
+- Changing paths of existing Datasets is no longer possible.
+- The API for the python client for reserving manual uploads (in case the python client has access to the underlying storage) has changed. WK now decides the storage paths.
```
Maybe add a hint, that updating wklibs is required?

MichaelBuessemeyer reviewed (commented):
First feedback part :D
Datastore + Tracingstore changes are still missing

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/models/datasource/DataSourceStatus.scala:
```diff
@@ -0,0 +1,11 @@
+
+object DataSourceStatus {
+  val unreported: String = "No longer available on datastore."
+  val deletedByUser: String = "Deleted by user."
```
Everything public should have a type annotation to speed up incremental compilation.

fm3 commented on app/controllers/DatasetController.scala:
```diff
@@ -39,14 +44,46 @@ case class DatasetUpdateParameters(
+object ReserveManualAttachmentUploadRequest {
+  implicit val jsonFormat: OFormat[ReserveManualAttachmentUploadRequest] =
+    Json.format[ReserveManualAttachmentUploadRequest]
+}
```
Yes, previously this request was sent to the datastore, where it also had side-effects on disk. But since it is now only in the DB, I moved it the request and its type definitions over here.

fm3 commented on app/controllers/DataStoreController.scala:
```diff
@@ -28,14 +28,8 @@ class DataStoreController @Inject()(dataStoreDAO: DataStoreDAO,
       (__ \ "publicUrl").read[String] and
       (__ \ "key").read[String] and
```
I’m not sure but I think maybe you misread the diff. This is only json parsing code and it wasn’t touched in this PR. The `dataStorePublicReads` was completely removed, because users are not supposed to edit datstores via the API at all.

fm3 commented on app/controllers/UserTokenController.scala:
```diff
@@ -198,10 +200,8 @@ class UserTokenController @Inject()(datasetDAO: DatasetDAO,
-                                  userBox: Box[User],
-                                  token: Option[String]): Fox[UserAccessAnswer] =
+  private def handleTracingAccess(tracingId: String, mode: AccessMode, userBox: Box[User], token: Option[String])(
+      implicit mp: MessagesProvider): Fox[UserAccessAnswer] =
```
That’s true, it is needed in more spots now, this is because it is used by usableDataSourceFor, which is used in many places now. You don’t need to pay special attention. Indeed, the compiler will emit an error if it’s missing and a warning if it’s superfluous. Since we elevate warnings to errors in the CI, we should be safe.

fm3 commented on app/controllers/WKRemoteDataStoreController.scala:
```diff
@@ -295,54 +296,19 @@ class WKRemoteDataStoreController @Inject()(
-                                                      abstractDataSource,
-                                                      isUsable = true)(GlobalAccessContext)
+          _ <- datasetDAO.findOne(datasetId)(GlobalAccessContext) ~> NOT_FOUND
+          _ <- datasetDAO.updateDataSource(datasetId, name, request.body.hashCode(), request.body, isUsable = true)(
```
Good catch! I found this problem too and am fixing it in https://github.com/scalableminds/webknossos/pull/8894. I’m not sure if that PR should be merged in here first or if we should keep them separate. Do you have a preference?

fm3 replied:
Ok let’s keep them separate for the moment to simplify the review at least. We can still consider what the best merging/rollback strategy will be when we get to that.

fm3 commented on app/models/dataset/Dataset.scala:
```diff
@@ -434,15 +430,12 @@ class DatasetDAO @Inject()(sqlClient: SqlClient, datasetLayerDAO: DatasetLayerDA
-  def doesDatasetDirectoryExistInOrganization(directoryName: String, organizationId: String)(
-      implicit ctx: DBAccessContext): Fox[Boolean] =
+  def doesDatasetNameExistInOrganization(datasetName: String, organizationId: String): Fox[Boolean] =
     for {
```
Normal dataset uploads now also use generated directoryNames, which include the objectId, so uniqueness checks should no longer be necessary.

fm3 commented on app/models/dataset/DatasetService.scala:
```diff
@@ -277,11 +256,112 @@ class DatasetService @Inject()(organizationDAO: OrganizationDAO,
+              e.dataFormat,
+              u.boundingBox,
+              e.elementClass,
+              e.mags,
```
Yes, adding mags would add new paths, which we don’t want to allow the user to define.

Again, if users really want to add mags later we might need to add something similar to reserveManualAttachmentUpload to enable it. But I think it is a rare use case.

fm3 commented on conf/evolutions/reversions/140-allows-manual-upload.sql:
```diff
@@ -0,0 +1,11 @@
+
+UPDATE webknossos.releaseInformation SET schemaVersion = 139;
+
+COMMIT TRANSACTION;
```
good catch!

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/storage/RemoteSourceDescriptorService.scala:
```diff
@@ -59,63 +62,43 @@ class RemoteSourceDescriptorService @Inject()(dSRemoteWebknossosClient: DSRemote
-  }
-
-  def resolveMagPath(datasetDir: Path, layerDir: Path, layerName: String, magLocator: MagLocator): URI =
+  def resolveMagPath(localDatasetDir: Path, layerDir: Path, layerName: String, magLocator: MagLocator): UPath =
```
Just a rather general question: Most of the method you are working with `Path`s and not `UPath`s while the  `UPath` with its / operation seems more convinient. So that might be a good thing to change :thinking:. But not strong opinion here as the `LocalUPath` uses a `Path` under the hood anyway

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/storage/RemoteSourceDescriptorService.scala:
```diff
@@ -59,63 +62,43 @@ class RemoteSourceDescriptorService @Inject()(dSRemoteWebknossosClient: DSRemote
-                               dataSourceId: DataSourceId,
-                               layerName: String,
-                               magLocator: MagLocator): Box[URI] = tryo {
+  private def resolveMagPath(baseDir: Path,
```
During first reading it wasn't so easy to see what the difference between this `resolveMagPath` and the one above is. Or more like, why are both needed. Maybe things might be a little clearer if one of the two methods gets a small renaming?

But not so super important imo

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/models/datasource/ElementClass.scala:
```diff
@@ -0,0 +1,168 @@
+
+  /* only used for segmentation layers, so only unsigned integers 8 16 32 64 */
+  private def maxSegmentIdValue(elementClass: ElementClass.Value): Long = elementClass match {
+    case ElementClass.uint8  => (1L << 8L) - 1
```
you added the -1 for correctness?

fm3 replied:
Yes, originally I just moved this code, but this problem was pointed out by coderabbit and I think it’s actually correct.

MichaelBuessemeyer commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/models/datasource/ElementClass.scala:
```diff
@@ -0,0 +1,168 @@
+    elementClass match {
+      case ElementClass.uint8  => Full(ElementClassProto.uint8)
+      case ElementClass.uint16 => Full(ElementClassProto.uint16)
+      case ElementClass.uint24 => Full(ElementClassProto.uint24)
```
And this was missing, correct?

fm3 replied:
Yes. I thought ElementClassProto doesn’t have uint24, since it doesn’t really make sense for VolumeTracingsLayers anyway, but apparently it does have it, so I think for completeness it should also be handled here.

MichaelBuessemeyer reviewed (commented):
Here is the final feedback part :D

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/helpers/PathSchemes.scala:
```diff
@@ -0,0 +1,12 @@
+
+  def isRemoteScheme(uriScheme: String): Boolean =
+    List(schemeS3, schemeHttps, schemeHttp, schemeGS).contains(uriScheme)
+}
```
> Why do you use explicit type annotations?

As stated above, every public member should have a type annotation to help optimize incremental compilation

> isRemoteScheme is marked as unused in my IDE

Well spotted, that’s right we don’t need it anymore. I removed it now. That also solves the unnecessary list creation.

> Why not use an enum?

Yeah that could work, but I”m not sure we want to reject parsing anything with an unknown scheme straightaway. I feel like that reduces our freedom to use UPath a little too much. But let’s reconsider this after some time and experience gathered with the UPath. I’ll add it to the list of potential follow-ups.

fm3 commented on app/models/dataset/DatasetService.scala:
```diff
@@ -85,39 +66,39 @@ class DatasetService @Inject()(organizationDAO: OrganizationDAO,
-      abstractDataSource = dataSource.copy(dataLayers = dataSource.dataLayers.map(AbstractDataLayer.from),
-                                           id = DataSourceId(datasetName, user._organization))
-      dataset <- createDataset(dataStore, newDatasetId, datasetName, abstractDataSource, isVirtual = true)
+      directoryName = generateDirectoryName(datasetName, newDatasetId)
```
Ok thanks, then I’d keep the new logic if it’s alright with you.

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/DataStoreConfig.scala:
```diff
@@ -21,7 +22,7 @@ class DataStoreConfig @Inject()(configuration: Configuration) extends ConfigRead
       val pingInterval: FiniteDuration = get[FiniteDuration]("datastore.webKnossos.pingInterval")
     }
-    val baseDirectory: String = get[String]("datastore.baseDirectory")
+    val baseDirectory: Path = Path.of(get[String]("datastore.baseDirectory")).toAbsolutePath
```
This setting is requried for datastore setups, datastore setup will fail if it is missing.

You’re right, it could directly use UPath, but I’m not so sure about that at this time. UPath does not extend nio.Path, so every usage that expects an nio.Path will need conversion back (or be adapted to handle UPath). Since we do know that this path will always be local, I tend to think it’s easier to keep using nio.Path here.

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/helpers/UPath.scala:
```diff
@@ -0,0 +1,185 @@
+  override def relativizedIn(potentialAncestor: UPath): UPath = this
+
+  override def hashCode(): Int =
+    new HashCodeBuilder(19, 29).append(scheme).append(segments).toHashCode
```
Good Idea! I’ll cache the whole hashCode in a lazy val, since those classes are all immutable anyway.

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/helpers/UPath.scala:
```diff
@@ -0,0 +1,185 @@
+}
+
+object UPath {
+  def separator: Char = '/'
```
Good catch, will do :)

fm3 commented on webknossos-datastore/app/com/scalableminds/webknossos/datastore/datavault/S3DataVault.scala:
```diff
@@ -106,7 +109,7 @@ class S3DataVault(s3AccessKeyCredential: Option[S3AccessKeyCredential],
       tc: TokenContext): Fox[(Array[Byte], Encoding.Value)] =
     for {
-      objectKey <- S3DataVault.objectKeyFromUri(path.toUri).toFox
+      objectKey <- S3DataVault.objectKeyFromUri(path.toRemoteUriUnsafe).toFox
```
Yes, see [comment above](https://github.com/scalableminds/webknossos/pull/8844#discussion_r2329613427)

fm3 commented:
@MichaelBuessemeyer please have another look at the latest changes :) I applied most of your feedback and commented on the rest.

I also added two more changes we did not talk about previously:
 - Added some code to ensure virtual datasets don’t get updated by the directory scanning in the datastore in case the dataset is virtual but still resides in the normal binaryData dir.
 - Changed the config for the path prefix to allow configuring multiple ones, and the client to select one. This is to support server infrastructure with different file systems mounts (Not the S3 case)

MichaelBuessemeyer reviewed (commented):
Ok, I just tried to explore a wkw dataset which is unsupported. Exploring a local zarr dataset worked instantly without any problems.

MichaelBuessemeyer commented:
Is there an open TODO for me? You need an approval from my side correct?

Is there anything that needs re-reviewing / testing as you made some changes?

fm3 commented:
Yes, the latest changes are:
 - Re-Add v10 reserveManualUpload, creating backwards compatibility for older libs versions. That means we can merge this PR as soon as it is approved :)
 - Renaming of the new APIs
 - added `reserveUploadToPathsForPreliminary` route, which is to be used (probably only) by the convert_to_wkw worker job. This allows to fill an “uploading” dataset, which the worker converts from e.g. tif to wkw.

I’d be happy if we could merge this in the coming week. Maybe you could have another quick look at the latest changes

MichaelBuessemeyer commented on app/models/job/Job.scala:
```diff
@@ -74,10 +74,15 @@ case class Job(
-          // New jobs return the directory name. Thus, the resulting link should be
-          returnValue.map { resultDatasetDirectoryName =>
-            s"/datasets/$organizationId/$resultDatasetDirectoryName/view"
+          // There exist jobs with dataset name as return value, some with directoryName, and newest with datasetId
+          // Construct links that work in either case.
+          returnValue.map { datasetIdentifier =>
+            ObjectId
+              .fromStringSync(datasetIdentifier)
+              .map { asObjectId =>
+                s"/datasets/$asObjectId/view"
+              }
+              .getOrElse(s"/datasets/$organizationId/$datasetIdentifier/view")
```
Neat, thanks for fixing :tada:

MichaelBuessemeyer commented on app/controllers/DatasetController.scala:
```diff
@@ -624,12 +634,24 @@ class DatasetController @Inject()(userService: UserService,
+      datasetId: ObjectId): Action[ReserveDatasetUploadToPathsForPreliminaryRequest] =
+    sil.SecuredAction.async(validateJson[ReserveDatasetUploadToPathsForPreliminaryRequest]) { implicit request =>
+      for {
+        dataset <- datasetDAO.findOne(datasetId) ?~> notFoundMessage(datasetId.toString) ~> NOT_FOUND
```
Or is it:
1. Call old reserverManualUpload
2. Upload to DS to S3
3. Inform wk where the DS is by calling this route?
I think that's wrong :thinking: ...

MichaelBuessemeyer reviewed (approved):
Awesome! Thanks a lot for your hard work! 

Let's give it a go :green_circle:
